src/main/java/net/neoforged/neoforge/transfer/handlers/wrappers/items/ItemContainerContentsResourceHandler.java

```diff
@@ -0,0 +1,169 @@
+    public void set(int index, ItemResource resource, int amount) {
+        ItemContainerContents contents = getContents();
+        contents.getStackInSlot(index).setCount(amount);
+        try (Transaction tx = TransactionManager.open(null)) {
+            setAndValidate(contents, amount, tx);
+            tx.commit();
+        }
+    }
+
+    @Override
+    public int insert(int index, ItemResource resource, int amount, TransactionContext context) {
+        if (ResourceHandlerUtil.isEmpty(resource, amount) || !isValid(index, resource)) return 0;
+        ItemContainerContents contents = getContents();
+        ItemStack stack = contents.getStackInSlot(index);
+        if (stack.isEmpty()) {
+            amount = Math.min(amount, resource.getMaxStackSize());
+
+            return setAndValidate(contents.with(index, resource, amount), amount, context);
+        } else if (resource.is(stack) && stack.getCount() < resource.getMaxStackSize()) {
+            int newAmount = Math.min(stack.getCount() + amount, resource.getMaxStackSize());
+            amount = newAmount - stack.getCount();
+//            if (action.isExecuting())
+            stack.grow(amount);
+            return setAndValidate(contents, amount, context);
+        }
+        return 0;
+    }
+
+    @Override
+    public int insert(ItemResource resource, int amount, TransactionContext context) {
+        if (ResourceHandlerUtil.isEmpty(resource, amount)) return 0;
+        ItemContainerContents contents = getContents();
+        int remaining = amount;
+        for (int i = 0; i < size; i++) {
+            ResourceStack<ItemResource> stack = contents.getStackInSlot(i).immutable();
+            if (stack.isEmpty() || !stack.resource().equals(resource) || stack.amount() >= resource.getMaxStackSize()) continue;
+            int toInsert = Math.min(remaining, resource.getMaxStackSize() - stack.amount());
+            contents = contents.with(i, resource, stack.amount() + toInsert);
+            remaining -= toInsert;
+        }
+        for (int i = 0; i < size; i++) {
+            ResourceStack<ItemResource> stack = contents.getStackInSlot(i).immutable();
+            if (!stack.isEmpty()) continue;
+            int toInsert = Math.min(remaining, resource.getMaxStackSize());
+            contents = contents.with(i, resource, toInsert);
+            remaining -= toInsert;
+            if (remaining <= 0) {
+                break;
+            }
+        }
+        return setAndValidate(contents, amount - remaining, context);
+    }
+
+    @Override
+    public int extract(int index, ItemResource resource, int amount, TransactionContext context) {
+        if (ResourceHandlerUtil.isEmpty(resource, amount)) return 0;
+        ItemContainerContents contents = getContents();
+        ResourceStack<ItemResource> stack = contents.getStackInSlot(index).immutable();
+        if (stack.isEmpty() || !stack.resource().equals(resource)) return 0;
```

Comment:
We already know the passed in resource isn't empty by here. This can just validate that the stack is not equal to the resource and then short-circuit. We don't need to explicitly check if the stack is empty.